### PR TITLE
Search with Elasticlunr, updated

### LINKED ISSUE #51: Have a search bar
Having a search bar would be very very useful. [rustbyexample](http://rustbyexample.com/) has one for example. I'm not sure how it would be implemented. I'm sure this has long been wanted by the [rust book](doc.rust-lang.org/book/).

Being able to search for programming keywords would actually make it very unusual and better than most too (outside the scope of a regular search bar though)

### PR DESCRIPTION
**Pull request by mattico**
@Phaiax's #472, manually rebased, plus a few improvements.

Major Changes from #472:
 - Breadcrumbs are made by adding `Chapter.parent_names`, rather than modifying the `BookItems` iterator
 - Markdown renderer renders footnotes, and also inserts spaces for certain tags to ensure words don't get merged together in the search results
 - `SearchDocument` removed, `render_item` renders directly into `elasticlunr::Index`
 - CSS modifications moved into stylus files
 - Removed JQuery from JS
 - Added integration test
 - Added aria attributes to search elements
 - Added search cargo feature:
   - Search index creation moved into module
   - Search JS moved into theme module for override/disable
 - HTML textification uses ammonia sanitizer

Closes #472 
Fixes #51

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -59,12 +62,13 @@ pretty_assertions = "0.4"
...
 
 [features]
-default = ["output", "watch", "serve"]
+default = ["output", "watch", "serve", "search"]
 debug = []
 output = []
 regenerate-css = []
 watch = ["notify", "time", "crossbeam"]
 serve = ["iron", "staticfile", "ws"]
+search = ["elasticlunr-rs"]
```
Comment (mattico) [inline Cargo.toml]: Forgot that I added this feature, it doesn't currently do anything. If desired, I can figure out how to extract the search functionality from hbs_renderer so it can be disabled.

**Comment by mattico** [comment]
@Michael-F-Bryan 

If there's anything I can do to make this easier to review, let me know. I'd be happy to break it into smaller commits.

No pressure, though. Wouldn't want a good maintainer to get burned out.

File: src/config.rs [hunk 2]
```diff
@@ -425,29 +425,84 @@ pub struct HtmlConfig {
...
         }
     }
 }
 
+/// Configuration of the search functionality of the HTML renderer.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(default, rename_all = "kebab-case")]
+pub struct Search {
+    /// Enable in browser searching. Default: `true` (if `search` feature is enabled).
+    pub enable: bool,
```
Comment (Michael-F-Bryan) [inline src/config.rs]: Is it necessary to have an `enabled` flag? I would have thought we can see search is enabled because a `output.html.search` is either present or not.

So an alternative to the `enabled` flag would be to change the `search` field in `HtmlConfig` to be `Option<Search>`... Thoughts?

File: src/config.rs [hunk 3]
```diff
@@ -425,29 +425,84 @@ pub struct HtmlConfig {
...
+    /// The path to the searcher to use. If not specified, will use the searcher code included in
+    /// MDBook. Default: `""`
+    pub searcher: PathBuf,
```
Comment (Michael-F-Bryan) [inline src/config.rs]: Hmm, I think we may end up having a similar issue to themes here. By adding this to the config file we're *theoretically* allowing people to swap in their own implementations, but because there is no well defined (and documented) interface to our `searcher`, people probably won't be able to make use of this knob in practice.

What would one gain by being able to use their own `searcher` implementations?

File: src/config.rs [hunk 4]
```diff
@@ -425,29 +425,84 @@ pub struct HtmlConfig {
...
+    /// Documents are split into smaller parts, seperated by headings. This defines, until which
+    /// level of heading documents should be split. Default: `3`. (`### This is a level 3 heading`)
+    pub heading_split_level: u8,
+}
+
+impl Default for Search {
+    fn default() -> Search {
+        // Please update the documentation of `Search` when changing values!
+        Search {
+            enable: cfg!(feature = "search"),
```
Comment (Michael-F-Bryan) [inline src/config.rs]: Nice use of the `cfg!()` macro! :+1:

File: src/renderer/html_handlebars/hbs_renderer.rs [hunk 5]
```diff
@@ -415,6 +413,14 @@ fn make_data(root: &Path, book: &Book, config: &Config, html_config: &HtmlConfig
                     json!("theme-tomorrow_night.js"));
     }
 
+    if html.search.enable {
+        if cfg!(feature = "search") {
+            data.insert("search_enabled".to_owned(), json!(true));
+        } else {
+            warn!("mdBook compiled without search support, ignoring `output.html.search.enable`");
```
Comment (Michael-F-Bryan) [inline src/renderer/html_handlebars/hbs_renderer.rs]: Good idea with adding a warning! Should we also say something like "please reinstall with `cargo install mdbook --force --features search` to use the search feature"?

File: src/renderer/html_handlebars/hbs_renderer.rs [hunk 6]
```diff
@@ -690,9 +647,9 @@ mod tests {
 
     #[test]
     fn anchor_generation() {
-        assert_eq!(id_from_content("## `--passes`: add more rustdoc passes"),
+        assert_eq!(utils::id_from_content("## `--passes`: add more rustdoc passes"),
```
Comment (Michael-F-Bryan) [inline src/renderer/html_handlebars/hbs_renderer.rs]: Now that the `id_from_content()` function has been moved to the `utils` module it may make sense to relocate these tests as well. I'm also curious to see how the ID generation will go when your header includes emojis, bold, or bits of code.

File: src/renderer/html_handlebars/search.rs [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+use book::BookItem;
+use config::Search;
+use errors::*;
+use renderer::RenderContext;
+use utils;
+use theme::searcher::Searcher;
+
+/// Creates all files required for search.
+pub fn create_files(ctx: &RenderContext) -> Result<()> {
+    let search_config = ctx.config.html_config().unwrap_or_default().search;
```
Comment (Michael-F-Bryan) [inline src/renderer/html_handlebars/search.rs]: Is there any chance we can pass in a `&HtmlConfig` here? It seems such a waste to be re-deserializing it when the caller should already have one lying around.

File: src/renderer/html_handlebars/search.rs [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+            | Event::End(Tag::Paragraph)
+            | Event::Start(Tag::Rule)
+            | Event::End(Tag::TableHead)
+            | Event::End(Tag::TableCell)
+            | Event::Start(Tag::List(_))
+            | Event::End(Tag::Item)
+            | Event::Start(Tag::Image(..))
+            | Event::End(Tag::Image(..))
+            | Event::SoftBreak
+            | Event::HardBreak => {
```
Comment (Michael-F-Bryan) [inline src/renderer/html_handlebars/search.rs]: This match statement looks like it was painful and error-prone to write out. Unfortunately I can't think of a better way of expressing this than what you've already got :disappointed:

File: src/theme/searcher/mod.rs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+pub static JS: &'static [u8] = include_bytes!("searcher.js");
+pub static MARK_JS: &'static [u8] = include_bytes!("mark.min.js");
+pub static ELASTICLUNR_JS: &'static [u8] = include_bytes!("elasticlunr.min.js");
+
+/// The `Searcher` struct should be used instead of the static variables because
+/// the `new()` method will look if the user has a searcher directory in their source folder and use
+/// the user's searcher instead of the default.
+///
+/// You should only use the static variables if you need the default searcher even if the
+/// user has specified another searcher.
```
Comment (Michael-F-Bryan) [inline src/theme/searcher/mod.rs]: It seems like a better way of expressing this warning would be to make the statics private and then implement `Default` for `Searcher` so people still have a way of accessing the defaults. 

Also, I don't think the doc-comment mentions what this struct is actually for... Although from context it's obviously a bag of bytes containing the JS assets, with a nice constructor for letting users provide their own versions.

File: src/utils/fs.rs [hunk 10]
```diff
@@ -16,6 +16,27 @@ pub fn file_to_string<P: AsRef<Path>>(path: P) -> Result<String> {
     Ok(content)
 }
 
+/// Changes path to use standard UNIX/URL separators
+pub fn normalize_path(path: &str) -> String {
```
Comment (Michael-F-Bryan) [inline src/utils/fs.rs]: This is just a thought, but would it make sense to use a `Path` or `Url` here to make it more obvious what these strings are?

File: src/utils/fs.rs [hunk 11]
```diff
@@ -16,6 +16,27 @@ pub fn file_to_string<P: AsRef<Path>>(path: P) -> Result<String> {
...
+    use std::path::is_separator;
+    path.chars()
+        .map(|ch| if is_separator(ch) { '/' } else { ch })
+        .collect::<String>()
+}
+
+/// Write the given data to a file, creating it first if necessary
+pub fn write_file<P: AsRef<Path>>(
```
Comment (Michael-F-Bryan) [inline src/utils/fs.rs]: **Thank you** for removing this from the HTML renderer! It actually started life as a helper method on `MDBook` and got moved out because it coupled the entire HTML renderer to `MDBook`'s internals. At the time I was refactoring a different part of the system, so didn't want to go off on a tangent by also fixing `write_file()`.

File: src/utils/mod.rs [hunk 12]
```diff
@@ -3,13 +3,60 @@
...
 
 use pulldown_cmark::{html, Event, Options, Parser, Tag, OPTION_ENABLE_FOOTNOTES,
                      OPTION_ENABLE_TABLES};
+
 use std::borrow::Cow;
 
 pub use self::string::{RangeArgument, take_lines};
 
+pub fn remove_html_tags<'a>(text: &'a str) -> Cow<'a, str> {
+    let regex = Regex::new(r"<[^>]*?>").unwrap();
```
Comment (Michael-F-Bryan) [inline src/utils/mod.rs]: Is it a good idea to use a regex here instead of something like a HTML sanitizer? I'm not 100% familiar with how the `pulldown-cmark` markdown renderer deals with angle brackets, but if we're not careful when calling it, we could accidentally clobber the wrong stuff (e.g. an innocent paragraph like "if \`x < 0\` and \`x > -5\`, blah blah blah").

Of course, that approach could also be overkill and 95% of the time the regex will work just fine...

File: src/renderer/html_handlebars/search.rs [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+fn make_doc_ref(anchor_base: &str, section_id: Option<String>) -> Cow<str> {
+    if let Some(id) = section_id {
+        format!("{}#{}", anchor_base, id).into()
+    } else {
+        anchor_base.into()
+    }
+}
+
+/// Renders markdown into flat unformatted text and adds it to the search index.
+fn render_item(index: &mut Index, search_config: &Search, item: &BookItem) -> Result<()> {
```
Comment (Michael-F-Bryan) [inline src/renderer/html_handlebars/search.rs]: Is it worth creating a bunch of tests for this so we can be sure the text gets stripped as intended?

File: src/renderer/html_handlebars/search.rs [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+        let doc_ref = make_doc_ref(&anchor_base, section_id);
+
+        index.add_doc(&doc_ref, &[&heading, &body, &breadcrumbs.join(" » ")]);
+    }
+
+    Ok(())
+}
+
+/// Converts the index and search options to a JSON string
+fn write_to_json(index: Index, search_config: &Search) -> Result<String> {
```
Comment (Michael-F-Bryan) [inline src/renderer/html_handlebars/search.rs]: Do you have access to any "known good" versions of the generated JSON? It's probably a good idea to add a test or two to make sure we're able to generate valid JSON, that way we can also prevent accidental regressions in the future if someone needs to come back and change things.

File: src/theme/searcher/mod.rs [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+pub struct Searcher {
+    pub js: Vec<u8>,
+    pub mark_js: Vec<u8>,
+    pub elasticlunr_js: Vec<u8>,
+}
+
+impl Searcher {
+    pub fn new(src: &Path) -> Self {
```
Comment (Michael-F-Bryan) [inline src/theme/searcher/mod.rs]: Should we create a test for this to make sure the file loading logic is sound and people are able to override the scripts?

File: src/config.rs (hunk 2 above)
Comment (mattico) [reply src/config.rs]: Agreed. Especially now that the JS is conditionally included, the flag is not very useful.

File: src/config.rs (hunk 3 above)
Comment (mattico) [reply src/config.rs]: I agree that the way it's setup is not ideal, I just copied the setup used for playpen_editor.

While we give no guarantees about the stability or interfaces for the searcher, the current system can be useful in practice for making small changes to search behavior (remove highlighting, change keyboard shortcuts, etc.). I'll agree, however, that this feature is unlikely to be used much if at all. In general I don't think many users will make significant theme changes out-of-tree until we provide a better way to do that.

So what change can we make to this PR? Here's what I'm thinking:

1. Remove the `editor` and `searcher` settings and the corresponding file override behavior.
2. Add `copy-js` settings for both. If set to false, the JS code for the feature will not be copied to the output directory. This is more flexible since users can then include whatever JS they want with `additional-js`, and it reduces the complexity of this little-used feature to almost zero.

File: src/config.rs (hunk 2 above)
Comment (mattico) [reply src/config.rs]: Hm, that would change search to be disabled by default. Is that your intention?

File: src/renderer/html_handlebars/search.rs (hunk 8 above)
Comment (mattico) [reply src/renderer/html_handlebars/search.rs]: It was trial-and-error, yes, but the good thing is that if it's wrong it doesn't matter that much. Now that I think about it the match could be replaced with `Event::Start(_) | Event::End(_) | Event::SoftBreak | Event::HardBreak` and the only difference would be a few more spurious spaces in the search result teasers. Hmm, and we could collapse multiple spaces down to one.

File: src/utils/fs.rs (hunk 10 above)
Comment (mattico) [reply src/utils/fs.rs]: The function operates on a string but `Path` needs to be converted into a string, which can throw an error. Both places this function is used already converted the path to a string. I don't think it's worth the trouble.

File: src/utils/mod.rs (hunk 12 above)
Comment (mattico) [reply src/utils/mod.rs]: Using a Regex on HTML does feel bad, but I didn't want to pull in a whole sanitizer. Angle brackets already have to be escaped in HTML blocks so it could clobber some data, but I blame the user. Worst-case scenario is that some pathological text inside of an HTML tag isn't in the search index.

File: src/renderer/html_handlebars/search.rs (hunk 14 above)
Comment (mattico) [reply src/renderer/html_handlebars/search.rs]: It's difficult to have any significant test for the entire search index. Tiny changes in the source material change the index drastically, and it's so large that it's difficult to determine if any change matters at all. If you have suggestions, I'm all ears.

File: src/config.rs (hunk 3 above)
Comment (Michael-F-Bryan) [reply src/config.rs]: I think that sounds quite reasonable :+1:

Besides, it's not hard to put back in if we find people actually want the feature.

File: src/config.rs (hunk 2 above)
Comment (Michael-F-Bryan) [reply src/config.rs]: I dunno, to be honest it's one of those features you'll always want for your document. 

It should probably be always turned on unless you manually deactivate the `search` feature flag when installing `mdbook` (e.g. when packaging for Debian).

File: src/renderer/html_handlebars/search.rs (hunk 14 above)
Comment (Michael-F-Bryan) [reply src/renderer/html_handlebars/search.rs]: Hmm, you make a good point... What I usually do is generate a small example and store that as a file, giving it a quick look through to make sure it seems alright. Then every time after that you can have a test which uses the same input and check the rendered document is the same as our known good copy.

I have a similar issue at work where it's not really possible to manually construct an object and check each of its elements are the same every time. For that I've got a `create_fixture()` function which looks somewhat like this:

```rust
/// Regenerates the test fixture (pre-nested list of cutlines).
///
/// # Note
///
/// You'll need to run `cargo test` manually here, if tests are being run via
/// `cargo watch` and you try to regenerate the test fixture we'll get midway
/// through writing out the fixture before `cargo watch` restarts the process.
/// Leaving a mangled fixture file.
const GENERATE_FIXTURE: bool = false;


// loads of tests


fn get_fixture() -> Vec<Cutline> {
    if cfg!(windows) && GENERATE_FIXTURE {
        let inputs = create_known_dummy_inputs();
        let thing = generate_output(&inputs).unwrap();

        let fixture_path =
            Path::new(env!("CARGO_MANIFEST_DIR")).join("tests/data/fixture.json");
        let mut f = ::std::fs::File::create(&fixture_path).unwrap();
        serde_json::to_writer_pretty(&mut f, &thing).unwrap();

        nested
    } else {
        let jason = include_str!("data/fixture.json");
        serde_json::from_str(jason).expect("Unable to deserialize the fixture")
    }
}
```

Comment (mattico) [reply src/renderer/html_handlebars/search.rs]: Ah that's a good solution! One of the issues I was running into is that rendering dummy_book is different when you're in a test (due to env vars?) so it's annoying to make fixtures.

**Comment by mattico** [comment]
I think that addresses everything, except for the JS dependency question. I would like to address that, I'm working out a solution right now. I think we should leave that for another PR, though.

**Comment by Phaiax** [comment]
I just wanted to say thank you for taking this over. :+1:

(I wished I had enough time for this but nope, got a new job and more...)

File: src/book/book.rs [hunk 16]
```diff
@@ -152,15 +152,20 @@ pub struct Chapter {
     pub sub_items: Vec<BookItem>,
     /// The chapter's location, relative to the `SUMMARY.md` file.
     pub path: PathBuf,
+    /// An ordered list of the names of each chapter above this one, in the hierarchy.
+    pub parent_names: Vec<String>,
```
Comment (Michael-F-Bryan) [inline src/book/book.rs]: Seeing that we need to manually store the parent names and keep them in sync makes me question whether it may be better to represent a `Book` as containing a "doubly-linked" tree of chapters, or even turn it into a graph structure. That way a `Chapter` can contain a pointer back to its parent and you can use that to walk up and down the levels of the tree. 

It would also hopefully simplify tracking section numbers and parsing. I might play around with this idea on a private branch and see how it goes.

File: src/config.rs [hunk 17]
```diff
@@ -425,29 +425,79 @@ pub struct HtmlConfig {
     pub livereload_url: Option<String>,
     /// Should section labels be rendered?
     pub no_section_label: bool,
+    /// Search settings. If `None`, the default will be used.
```
Comment (Michael-F-Bryan) [inline src/config.rs]: > If `None`, the default will be used.

I believe this is already sorted out by the `#[serde(default)]` annotation on `HtmlConfig`. That annotation tells serde to use `Search::default()` if it can't deserialize the settings.

To me, when I see `Option<Search>` I'd think this indicates that searching is optional and when no `Search` is found in my `book.toml` this means my book won't have a search bar... Am I interpreting it wrong, or is this comment out of sync with the implementation?

File: src/config.rs [hunk 18]
```diff
@@ -425,29 +425,79 @@ pub struct HtmlConfig {
...
-    /// The path to the editor to use. Defaults to the [Ace Editor].
-    /// 
-    /// [Ace Editor]: https://ace.c9.io/
-    pub editor: PathBuf,
-    /// Should playpen snippets be editable? Defaults to `false`.
+    /// Should playpen snippets be editable? Default: `false`.
     pub editable: bool,
+    /// Copy JavaScript files for the editor to the output directory?
+    /// Default: `true`.
+    pub copy_js: bool,
```
Comment (Michael-F-Bryan) [inline src/config.rs]: This probably belongs in its own PR. Changing how the playpen works and whether we copy across JavaScript files is a non-trivial change.

File: src/renderer/html_handlebars/hbs_renderer.rs [hunk 19]
```diff
@@ -307,15 +296,20 @@ impl Renderer for HtmlHandlebars {
         fs::create_dir_all(&destination)
             .chain_err(|| "Unexpected error when constructing destination path")?;
 
-        for (i, item) in book.iter().enumerate() {
+        let mut depthfirstiterator = book.iter();
+        let mut is_index = true;
+        while let Some(item) = depthfirstiterator.next() {
```
Comment (Michael-F-Bryan) [inline src/renderer/html_handlebars/hbs_renderer.rs]: Shouldn't this just be a plain old `for` loop?

File: tests/rendered_output.rs [hunk 20]
```diff
@@ -339,3 +339,101 @@ fn book_with_a_reserved_filename_does_not_build() {
...
+
+    // So you've broken the test. If you changed dummy_book, it's probably
+    // safe to regenerate the fixture. If you haven't then make sure that the
+    // search index still works. Run `cargo run -- serve tests/dummy_book`
+    // and try some searches. Are you getting results? Do the teasers look OK?
+    // Are there new errors in the JS console?
+    //
+    // If you're pretty sure you haven't broken anything, change `GENERATE_FIXTURE`
+    // above to `true`, and run `cargo test` to generate a new fixture. Then
+    // change it back to `false`.
```
Comment (Michael-F-Bryan) [inline tests/rendered_output.rs]: I got a nice laugh out of this block of comments, yet it's still surprisingly useful for people who see it because of a failing test. Thank you!

**Comment by Michael-F-Bryan** [comment]
I just tried this out on my laptop. One thing I've noticed is the search index is only available when running mdbook with `mdbook serve` which is a bit of a pain when testing with `mdbook build --open` locally. 

The browser console shows the following error.

```
searcher.js:260 Failed to load file:///home/michael/Documents/mdBook/book-example/book/searchindex.json: Cross origin requests are only supported for protocol schemes: http, data, chrome, chrome-extension, https.
init @ searcher.js:260
```

After a brief search it seems like this is a [Chrome-specific issue][1] and there isn't really any way around it other than invoking chrome from the command-line with the `--allow-file-access-from-files` flag. @sorin-davidoi, have you encountered this before?

[1]: http://ahmed.amayem.com/solution-to-chrome-cross-origin-requests-are-only-supported-for-protocol-schemes-http-data-chrome-chrome-extension-https-chrome-extension-resource/

File: src/renderer/html_handlebars/hbs_renderer.rs (hunk 19 above)
Comment (mattico) [reply src/renderer/html_handlebars/hbs_renderer.rs]: Yeah, I'm not sure why this was changed. This is from the original PR.

**Comment by sorin-davidoi** [comment]
Don't think the issue is Chrome-specific - you just can't read arbitrary files from the filesystem. A (nasty) workaround could be to make a `searchindex.js` file which contains the entire JSON and assigns it to `window` - you can then add the file as a script tag, wait for it to load and execute, and proceed with reading the JSON from `window`.

File: src/config.rs (hunk 17 above)
Comment (mattico) [reply src/config.rs]: > I believe this is already sorted out by the #[serde(default)] annotation on HtmlConfig. That annotation tells serde to use Search::default() if it can't deserialize the settings.

The default it'll use is `None`, since that's how `Default::default::<Option>()` is implemented: https://doc.rust-lang.org/src/core/option.rs.html#900-906, and we're using `#[derive(Default)]`. The html renderer will then manually use `unwrap_or_default` to get the default configuration. The reason to do this is so we can print the warning message if the user is using mdbook compiled without `feature = "search"`, but does have an `[output.html.search]` table in their `book.toml`.

File: src/config.rs (hunk 18 above)
Comment (mattico) [reply src/config.rs]: I suppose. Should I revert the changes to `editor`, but leave the `output.html.search.copy-js` setting?

**Comment by mattico** [comment]
I looked into HTML sanitization. `ammonia` isn't that big of a dependency, and it could work with [a small change](https://github.com/notriddle/ammonia/issues/89).

File: src/config.rs (hunk 17 above)
Comment (mattico) [reply src/config.rs]: It may be less confusing overall to just bring back the `[output.html.search.enabled]` field.

**Comment by mattico** [comment]
Probably wait for https://github.com/notriddle/ammonia/pull/90 now.

File: src/config.rs (hunk 18 above)
Comment (Michael-F-Bryan) [reply src/config.rs]: Mmm, don't worry about it for now, it's probably not worth the effort. As long as it's noted in the PR description it should be fine.

**Comment by Michael-F-Bryan** [comment]
@mattico, I just looked through the PR again and I'm pretty happy with it :+1:

On a side note, it looks like `html5ever` emits a bunch of `unimplemented` warnings when generating the search index, do you know what may be causing them? The `stop_parsing` operation sounds like more of an optimisation to stop parsing HTML early so I doubt it's actually going to have any negative impact on `mdbook`, but the user experience isn't the best. 

We can always hack around this by updating the `init_logging()` function in `src/bin/mdbook.rs` to use `builder.filter(Some("mdbook"), LevelFilter::Info)` when the `RUST_LOG` environment variable isn't set, although I'd be interested to see if there's a way to fix the underlying `html5ever` issue.

For posterity's sake, this is what I see when running `mdbook build` locally:

```text
$ cargo run -- build --open book-example                                     
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running `target/debug/mdbook build --open book-example`
2018-03-05 16:15:36 [INFO] (mdbook::book): Book building has started
2018-03-05 16:15:36 [INFO] (mdbook::book): Running the html backend
2018-03-05 16:15:37 [WARN] (html5ever::tree_builder): stop_parsing not implemented, full speed ahead!
2018-03-05 16:15:37 [WARN] (html5ever::tree_builder): stop_parsing not implemented, full speed ahead!
2018-03-05 16:15:37 [WARN] (html5ever::tree_builder): stop_parsing not implemented, full speed ahead!
2018-03-05 16:15:37 [WARN] (html5ever::tree_builder): stop_parsing not implemented, full speed ahead!
2018-03-05 16:15:37 [WARN] (html5ever::tree_builder): stop_parsing not implemented, full speed ahead!
2018-03-05 16:15:37 [WARN] (html5ever::tree_builder): stop_parsing not implemented, full speed ahead!
2018-03-05 16:15:37 [WARN] (html5ever::tree_builder): stop_parsing not implemented, full speed ahead!
2018-03-05 16:15:37 [WARN] (html5ever::tree_builder): stop_parsing not implemented, full speed ahead!
Created new window in existing browser session.
```

Other than the annoying `html5ever` warnings, I think this is about ready to merge.

**Comment by mattico** [comment]
Ah, should've run more than cargo test. I'll look into it.

**Comment by mattico** [comment]
Fixed.

**Comment by Michael-F-Bryan** [comment]
And merged :tada:

**Comment by mattico** [comment]
4 months ago I looked at the search issue and thought "how hard can it be"?. 🎉🎊🎖️📯

**Comment by Michael-F-Bryan** [comment]
I've been using `master` locally for a couple days now and can't see any issues with search, so I'll probably to do another release either tomorrow or the day after.

Should I also make a post to [/r/rust]? Is there anything in particular you'd like me to say about `mdbook`'s new search functionality?

[/r/rust]: https://www.reddit.com/r/rust/

**Comment by mattico** [comment]
> Should I also make a post to /r/rust?

That's a great idea! People should know to rebuild their docs.

> Is there anything in particular you'd like me to say about mdbook's new search functionality?

Hmm. I'd mention that if anyone notices something which won't show up in search results that they should file an issue on here. There's possibly still some bugs in the index preprocessing or elasticlunr-rs.

If you're going to have a "thanks to" section make sure to mention phaiax for writing most of this ^^ and notriddle for maintaining ammonia.

There are a lot of questions on /r/rust like "what's a project I can contribute to to learn Rust". I think _mdbook_ works well for that, the code is easy to understand and its fun to see the results in the browser. On the other hand sometimes working on it is more HTML than Rust, and a lot of the stuff that needs to be worked on is more architecture than self-contained improvements.

**Comment by Michael-F-Bryan** [comment]
@mattico I think he have a [small CSS issue](https://rust-lang-nursery.github.io/mdBook/?search=command) with the search results 😜 
 
image

Do you know what may have caused this? I'm using Firefox on Windows, but I remember testing the search feature before our `0.1.4` release (Chrome on Arch Linux) and the search results would get inserted above the page contents just fine.

@sorin-davidoi, this should be a pretty trivial CSS fix, shouldn't it?
